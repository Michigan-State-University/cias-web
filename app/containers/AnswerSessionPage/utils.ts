import dayjs from 'dayjs';
import pickBy from 'lodash/pickBy';
import isNil from 'lodash/isNil';

import { TlfbConfigBody } from 'models/Question';
import { CalendarData } from 'models/Tlfb';
import { CamelToSnake } from 'global/types/camelToSnake';
import { fullDayToYearFormatter } from 'utils/formatters';

export const getSessionMapUserPreviewUrl = (userSessionId: string): string => {
  const {
    location: { pathname },
  } = window;
  const previewIndex = pathname.indexOf('preview');
  const sessionUrl = pathname.slice(0, previewIndex - 1);
  return `${sessionUrl}/map?userSessionId=${userSessionId}`;
};

export const getNextSessionUrl = (sessionId: string): string => {
  const {
    location: { pathname },
  } = window;

  return pathname.replace(/\/sessions\/.*/, `/sessions/${sessionId}/fill`);
};

export const getBackToModulesUrl = (): string => {
  const {
    location: { pathname },
  } = window;

  return pathname.replace(/\/sessions\/.*/, '/invite');
};

const generateGridCellId = (prefix: string, index: number) =>
  `${prefix}-${index}`;

export const generateGridAnswerId = (index: number) =>
  `${generateGridCellId('answer', index)}`;
export const generateGridQuestionId = (index: number) =>
  `${generateGridCellId('question', index)}`;

const getOldestDate = (dates: CalendarData) =>
  dayjs.min(
    Object.keys(dates).map((dayString) =>
      dayjs(dayString, fullDayToYearFormatter),
    ),
  );

export const getCalendarMetadata = (
  tlfbConfig: CamelToSnake<TlfbConfigBody>,
  calendarData: CalendarData,
) => {
  const {
    data: [
      {
        payload: { days_count: daysCount },
      },
    ],
  } = tlfbConfig;
  const yesterday = dayjs().subtract(1, 'day');
  const oldestFilledDate = getOldestDate(calendarData);
  const datesWithAnswers = pickBy(calendarData, (date) => !isNil(date.answer));
  const oldestFilledAnswerDate = getOldestDate(datesWithAnswers);
  const oldestAllowedDate = dayjs().subtract(+(daysCount || 1), 'day');
  const isEveryAnswerFilled =
    oldestFilledAnswerDate?.isSame(oldestAllowedDate, 'day') ?? false;

  return {
    yesterday,
    oldestAllowedDate,
    oldestFilledDate,
    oldestFilledAnswerDate,
    isEveryAnswerFilled,
  };
};
