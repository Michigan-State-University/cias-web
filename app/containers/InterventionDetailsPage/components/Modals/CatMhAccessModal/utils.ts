import { Intervention } from 'models/Intervention';

import { ModalUIData } from './types';

export const getTestsLeft = (
  catMhPool: Nullable<number>,
  createdCatMhSessionCount: Nullable<number>,
) => Math.max(0, (catMhPool ?? 0) - (createdCatMhSessionCount ?? 0));

export const mapInterventionToModalData = (
  intervention: Intervention,
): ModalUIData => {
  const {
    catMhPool,
    licenseType,
    catMhApplicationId,
    catMhOrganizationId,
    isAccessRevoked,
    createdCatMhSessionCount,
  } = intervention;

  return {
    testNumber: catMhPool ?? 0,
    testsLeft: getTestsLeft(catMhPool, createdCatMhSessionCount),
    licenseType,
    applicationId: catMhApplicationId ?? '',
    organizationId: catMhOrganizationId ?? undefined,
    isAccessRevoked,
  };
};

export const mapModalDataToIntervention = (
  modalData: ModalUIData,
): Partial<Intervention> => {
  const {
    testNumber,
    licenseType,
    applicationId,
    organizationId,
    isAccessRevoked,
  } = modalData;

  return {
    catMhPool: testNumber ?? 0,
    licenseType,
    catMhApplicationId: applicationId,
    catMhOrganizationId: organizationId,
    isAccessRevoked,
  };
};
