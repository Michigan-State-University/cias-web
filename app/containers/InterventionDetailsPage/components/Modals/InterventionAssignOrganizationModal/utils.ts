import { Organization } from 'models/Organization';

import { SelectOption } from 'components/Select/types';

export const organizationSelectOptionFormatter = ({
  id: value,
  name: label,
}: Organization): SelectOption<string> => ({
  value,
  label,
});
