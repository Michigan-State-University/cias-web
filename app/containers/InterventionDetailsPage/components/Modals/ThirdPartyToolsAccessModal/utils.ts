import { InterventionDto } from 'models/Intervention';

import { ModalUIData } from './types';

export const mapInterventionToModalData = (
  intervention: InterventionDto,
): ModalUIData => {
  const {
    catMhPool,
    licenseType,
    catMhApplicationId,
    catMhOrganizationId,
    isAccessRevoked,
    createdCatMhSessionCount,
    hfhsAccess,
  } = intervention;

  return {
    testNumber: catMhPool ?? 0,
    currentTestNumber: catMhPool - createdCatMhSessionCount ?? 0,
    licenseType,
    applicationId: catMhApplicationId ?? '',
    organizationId: catMhOrganizationId ?? undefined,
    isAccessRevoked,
    hfhsAccess,
  };
};

export const mapModalDataToIntervention = (
  modalData: ModalUIData,
): Partial<InterventionDto> => {
  const {
    testNumber,
    licenseType,
    applicationId,
    organizationId,
    isAccessRevoked,
    hfhsAccess,
  } = modalData;

  return {
    catMhPool: testNumber ?? 0,
    licenseType,
    catMhApplicationId: applicationId,
    catMhOrganizationId: organizationId,
    isAccessRevoked,
    hfhsAccess,
  };
};
