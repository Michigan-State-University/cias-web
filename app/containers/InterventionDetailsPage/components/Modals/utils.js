export const organizationSelectOptionFormatter = ({
  id: value,
  name: label,
}) => ({
  value,
  label,
});
