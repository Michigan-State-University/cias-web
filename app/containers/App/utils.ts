import { matchPath } from 'react-router';

import { PasswordAuthenticatedRoles, Roles } from 'models/User/RolesManager';

import { arraysOverlap } from 'utils/arrayUtils';

import { ALL_ROUTE_PATHS, RoutePath } from 'global/constants';

import { OMIT_FETCH_SELF_DETAILS_PATHS } from './constants';

export const shouldFetchSelfDetailsOnPath = (pathname: string) => {
  const match = matchPath(pathname, { path: ALL_ROUTE_PATHS, exact: true });
  return (
    match && !OMIT_FETCH_SELF_DETAILS_PATHS.includes(match.path as RoutePath)
  );
};

export const shouldFetchSelfDetailsByUserRoles = (userRoles: Roles[]) =>
  arraysOverlap(userRoles, PasswordAuthenticatedRoles);
