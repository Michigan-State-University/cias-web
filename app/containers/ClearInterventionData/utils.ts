import * as Yup from 'yup';
import { nonNegativeIntegerValidationSchema } from 'utils/validators';

export const createClearConversationDataFormValidationSchema = () =>
  Yup.object().shape({
    delay: nonNegativeIntegerValidationSchema,
  });
