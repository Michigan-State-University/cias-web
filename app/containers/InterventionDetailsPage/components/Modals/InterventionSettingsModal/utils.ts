import * as Yup from 'yup';

import { Intervention, InterventionType } from 'models/Intervention';
import { ShortLinkData } from 'models/ShortLink';

import { jsonApiToArray } from 'utils/jsonApiMapper';
import { LanguageSelectOption } from 'utils/formatters';
import {
  requiredValidationSchema,
  unreservedURLCharactersSchema,
} from 'utils/validators';

import {
  InterventionSettingsFormValues,
  GetShortLinksResponse,
  ShortLinksData,
} from './types';

// TODO validate duplicates
export const InterventionSettingsFormValidationSchema = Yup.object().shape({
  links: Yup.array().of(
    Yup.object({
      selected: Yup.boolean(),
      name: Yup.string().when('selected', {
        is: (selected) => selected,
        then: unreservedURLCharactersSchema.concat(requiredValidationSchema),
      }),
      healthClinicId: Yup.string(),
    }),
  ),
});

export const getShortLinksDataParser = (
  data: GetShortLinksResponse,
): ShortLinksData => ({
  shortLinks: jsonApiToArray(data, 'shortLink'),
  healthClinics: jsonApiToArray(
    { data: data.health_clinics },
    'simpleHealthClinic',
  ),
});

export const mapShortLinksToFormValues = (
  shortLinksData: Nullable<ShortLinksData>,
  inOrganization: boolean,
): InterventionSettingsFormValues['links'] => {
  if (!inOrganization) {
    const links = shortLinksData?.shortLinks?.map(({ name }) => ({
      selected: !!name,
      name: name ?? '',
    }));
    return links?.length ? links : [{ selected: false, name: '' }];
  }

  return (
    shortLinksData?.healthClinics?.map(({ id }) => {
      const { name } =
        shortLinksData?.shortLinks?.find(
          ({ healthClinicId }) => healthClinicId === id,
        ) ?? {};
      return {
        selected: !!name,
        name: name ?? '',
        healthClinicId: id,
      };
    }) ?? []
  );
};

export const mapFormValuesToShortLinks = (
  links: InterventionSettingsFormValues['links'],
): ShortLinkData[] =>
  links
    .filter(({ selected, name }) => selected && name)
    .map(({ name, healthClinicId }) => ({ name: name!, healthClinicId }));

export const mapLanguageToInterventionChanges = ({
  value,
  label,
  googleLanguageId,
}: LanguageSelectOption): Pick<
  Intervention,
  'languageCode' | 'languageName' | 'googleLanguageId'
> => ({
  languageCode: value,
  languageName: label,
  googleLanguageId: +googleLanguageId,
});

export const getPlaceholderBase = (
  interventionId: string,
  interventionType: InterventionType,
  firstSessionId: Nullable<string>,
) => {
  const base = `${process.env.WEB_URL}/interventions/${interventionId}`;

  switch (interventionType) {
    case InterventionType.DEFAULT:
      if (!firstSessionId) return '';
      return `${base}/sessions/${firstSessionId}/fill`;
    default:
      return `${base}/invite`;
  }
};
