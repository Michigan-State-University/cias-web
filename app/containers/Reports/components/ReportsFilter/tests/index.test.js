/**
 *
 * Tests for ReportFilter
 *
 */

import React from 'react';
import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { DEFAULT_LOCALE } from 'i18n';
import { IntlProvider } from 'react-intl';
import { formatMessage } from 'utils/intlOutsideReact';
import {
  filterOptions,
  PARTICIPANTS,
  SORT_BY_LATEST,
} from 'global/reducers/generatedReports/constants';

import { ReportsFilter } from '../index';

describe('<ReportsFilter />', () => {
  const filterOpts = filterOptions(false);
  const defaultProps = {
    formatMessage,
    filterOptions: filterOpts,
    activeFilters: [PARTICIPANTS],
    activeSort: SORT_BY_LATEST,
  };
  it('Expect to not log errors in console', () => {
    const spy = jest.spyOn(global.console, 'error');
    render(
      <MemoryRouter>
        <IntlProvider locale={DEFAULT_LOCALE}>
          <ReportsFilter {...defaultProps} />
        </IntlProvider>
      </MemoryRouter>,
    );
    expect(spy).not.toHaveBeenCalled();
  });

  it('Should render and match the snapshot', () => {
    const {
      container: { firstChild },
    } = render(
      <MemoryRouter>
        <IntlProvider locale={DEFAULT_LOCALE}>
          <ReportsFilter {...defaultProps} />
        </IntlProvider>
      </MemoryRouter>,
    );
    expect(firstChild).toMatchSnapshot();
  });
});
