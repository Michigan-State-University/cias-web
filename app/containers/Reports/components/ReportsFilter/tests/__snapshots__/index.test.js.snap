// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ReportsFilter /> Should render and match the snapshot 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  height: auto;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  justify-items: space-between;
  -webkit-align-content: center;
  -ms-flex-line-pack: center;
  align-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-top: 24px;
  margin-bottom: 24px;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  height: auto;
  -webkit-align-content: center;
  -ms-flex-line-pack: center;
  align-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  height: auto;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  justify-items: space-between;
  -webkit-align-content: center;
  -ms-flex-line-pack: center;
  align-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c11 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  height: auto;
  min-width: 150px;
}

.c2 {
  font-family: DM Sans;
  font-size: 13px;
  line-height: 130%;
  margin-right: 15px;
  font-weight: 700;
  color: #2F3850;
}

.c10 {
  font-family: DM Sans;
  font-size: 13px;
  line-height: 130%;
  margin-right: 15px;
  font-size: 15px;
  color: #2F3850;
}

.c13 {
  font-family: DM Sans;
  font-size: 13px;
  line-height: 130%;
  font-weight: 700;
  color: rgba(183,39,234,1);
  cursor: pointer;
  color: rgba(183,39,234,1);
  cursor: pointer;
}

.c12 {
  width: auto;
  height: auto;
  display: block;
  border-radius: 5px;
  position: relative;
  color: unset;
}

.c14 {
  width: auto;
  height: auto;
  margin-left: 8px;
  cursor: pointer;
  -webkit-transition: -webkit-transform 0.2s;
  -webkit-transition: transform 0.2s;
  transition: transform 0.2s;
}

.c5 {
  display: none;
}

.c9 {
  margin-right: 5px;
  border-radius: 5px;
}

.c4:disabled + label > .c8 {
  opacity: 0.5;
  cursor: not-allowed;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
}

.c7:hover {
  cursor: pointer;
}

.c4:disabled + .c6 {
  opacity: 0.5;
  cursor: not-allowed;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <p
      class="c2"
      font-weight="bold"
    >
      Show reports generated for:
    </p>
    <div
      class="c1"
    >
      <div
        class="c3"
      >
        <input
          checked=""
          class="c4 c5"
          id="participant"
          type="checkbox"
          width="20"
        />
        <label
          class="c6 c7"
          for="participant"
        >
          <img
            class="c8 c9"
            src="IMAGE_MOCK"
          />
          <p
            class="c10"
            font-size="15"
          >
            Participants
          </p>
        </label>
      </div>
    </div>
    <div
      class="c1"
    >
      <div
        class="c3"
      >
        <input
          class="c4 c5"
          id="third_party"
          type="checkbox"
          width="20"
        />
        <label
          class="c6 c7"
          for="third_party"
        >
          <img
            class="c8 c9"
            src="IMAGE_MOCK"
          />
          <p
            class="c10"
            font-size="15"
          >
            3rd Party Users
          </p>
        </label>
      </div>
    </div>
  </div>
  <div
    class="c11"
  >
    <div
      class="c12"
    >
      <div
        class="c1"
        role="button"
      >
        <p
          class="c13"
          color="#B727EA"
          font-weight="bold"
        >
          Sort by Oldest
        </p>
        <img
          alt="arrow"
          class="c14"
          src="IMAGE_MOCK"
          transform=""
        />
      </div>
    </div>
  </div>
</div>
`;
